### PR TITLE
firewall: Explain what is role of interface when adding new zone

### LINKED ISSUE #12307: firewall: a little unclear what the role of selecting interface is
Cockpit version: 197
OS: Fedora Server 30
Page: Network > Firewall

In the usability test we ran now in July with 9 participants, it became apparent that it's not always super clear what adding an interface to a zone means.
Only one 1 out of 9 participants ran into this, but it got marked with Severity 3, so I thought I might as well file it as an issue.

**Quote from participant 2:** _Is it exclusive to that interface in that zone?_

Frequency: 11.11%
Severity: 3/5

### PR DESCRIPTION
**Pull request by marusak**
Fixes #12307

Screenshot from 2022-03-09 12-04-58

### CODE REVIEW

**Review by garrett — CHANGES_REQUESTED** [review]
We probably need to adjust some wording here. Otherwise, it's a good fixup.

The interface list is completely unorganized here and needs work... but in another PR. The checkboxes should be aligned, more like this:

image

(But, again: Another PR for the alignment. It's just glaringly obvious in the screenshot.)

File: pkg/networkmanager/firewall.jsx
Comment (marusak) [reply pkg/networkmanager/firewall.jsx:773]: > Also, this implies incoming and outgoing traffic, which isn't the case.. right?

I am not certain. Surely controlling incoming traffic is the most common usecase but I think that you can block also outgoing traffic with firewall. Not sure though if you can specify it in zones/services.

Comment (garrett) [reply pkg/networkmanager/firewall.jsx:773]: > I am not certain. Surely controlling incoming traffic is the most common usecase but I think that you can block also outgoing traffic with firewall. Not sure though if you can specify it in zones/services.

You cannot block outgoing traffic with Cockpit's firewall. It only handles incoming traffic.

I talked with @jelly about this and recently wrote current thoughts on how to handle it @ https://github.com/cockpit-project/cockpit/issues/9047#issuecomment-1026013963 with a bit of it covered in the mockups on the comment right before it @ https://github.com/cockpit-project/cockpit/issues/9047#issuecomment-1026005844.

(I did think a lot about it and wrote things about outgoing as well as incoming a long time ago, and I think I even did some mockups for it back then too — but the firewall page has changed since the initial drafts, so I needed to redo it for the current implementation of the firewall page.)

But the reason why we only block incoming traffic with the firewall page is that it's what's wanted on a server almost all of the time, as the software running on a server is almost always a few carefully chosen or custom written apps. Plus, it would mean we need someone to implement it. (We still don't have everything implemented that was designed and mocked up, even for just handling incoming traffic.)

Blocking outgoing is more useful for workstations, where there's a lot of software written by many different authors and so many are doing network connections these days... but Cockpit's for servers, not workstations; so we haven't implemented it yet (as it wouldn't be so useful in most cases).

If we say "all traffic", then we're implying that the firewall actually affects all traffic, including outgoing... and this isn't true.

Adding a zone on the system _may_ affect some outgoing traffic if there are additional rules in the zone for it, but we don't provide anything in the UI to actually show or add to or manage that at the moment.

Comment (garrett) [reply pkg/networkmanager/firewall.jsx:773]: To unstick this and move it forward, we could just drop the "all" and be intentionally vague (instead of trying to be specific):

```suggestion
                                             <FlexItem>{_("Traffic passing through selected interfaces will respect this zone's settings.")}</FlexItem>
                                             <FlexItem>{_("Each interface can be assigned only to one zone.")}</FlexItem>
```

It's still a little problematic, implying all but not outright saying it.

Once we land https://github.com/cockpit-project/cockpit/pull/17084 with some fixes, it should be clearer that the firewall is only for incoming traffic (for now at least).

Comment (marusak) [reply pkg/networkmanager/firewall.jsx]: Changed it to `Traffic passing through...` - as it is correct (if the zone settings don't do anything about outgoing, it still respects these settings :)) and not too specific in the case that some zone or included service would actually affect outgoing traffic.

File: pkg/networkmanager/firewall.jsx
Comment (garrett) [reply pkg/networkmanager/firewall.jsx:773]: @allisonkarlitskaya suggests:

```suggestion
                                             <FlexItem>{_("An interface cannot belong to multiple zones.")}</FlexItem>
```

Let's go with this.

**Review by garrett — CHANGES_REQUESTED** [review]
I've added some suggestions, including @allisonkarlitskaya's.

I don't like to have a negative statement here instead of a positive (cannot vs. can) — but it is clearer, so let's make an exception here.

WDYT, @marusak?

**Comment by garrett** [comment]
Once we change the interface text, I'm happy. We should merge this in after the change. :+1:

File: pkg/networkmanager/firewall.jsx
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx:768]: In*t*erfaces. Human visible and translated, so let's not land typos here.

File: pkg/networkmanager/firewall.jsx
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx:773]: I found "to only one zone" easier to understand, but no strong opinion. As this was vetted by Garrett and Lis already, fine for me.

Again, it feels a bit pointless to me to put that into a help text. It only aggravates the uncertainty of the user -- "So do I need to make sure that it's unique here? Why doesn't the UI enforce that?" *Do* we enforce that in firewalld/the UI actually? I should not be able to select the same interface in more than one zone, and then we wouldn't need to explain it either.

File: pkg/networkmanager/firewall.jsx
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx:772]: Just making sure, you are explicitly ignoring the incoming vs. outgoing question here, right?

Honestly, I'm not sure what this sentence is trying to tell me, it feels rather tautological -- if I add an interface to a zone, then yes, I expect it to follow the rules of that zone.

It feels like it should explain what a "zone" is (essentially, a collection of interfaces and binding them to rules)? Perhaps

> This zone consists of the selected interfaces

But still, I find it rather hard to imagine why *this* part would be the confusion. If I look at that dialog, the big question mark that I have is what on earth a "Trust level" is. I.e. what kind of influence/behaviour change does it have when I change "public" to "external"? (It's an honest question -- I have no idea)

**Review by martinpitt — CHANGES_REQUESTED** [review]
Honestly, to me this doesn't make matters any better. ISTM that we need to redesign this more thoroughly. But it's not a strong veto, other than the typo.

File: pkg/networkmanager/firewall.jsx
Comment (marusak) [reply pkg/networkmanager/firewall.jsx:773]: You cannot select interface from UI that is already assigned to different zone.

**Comment by marusak** [comment]
Ok, I seems that this change is not really necessary and does not provide anything useful. For someone who has no idea what zone and interface are this won't help and for someone who knows what they are doing this is useless. I am closing this as well as the original issue. The original issue also states it affected only 1 person out of 9.

File: pkg/networkmanager/firewall.jsx
Comment (martinpitt) [reply pkg/networkmanager/firewall.jsx:773]: *phew*, that's good. Thanks!

**Comment by martinpitt** [comment]
Could we add some help text for the trust level perhaps?

**Pull request closed without merging**
